src/ifcwrap-csharp/IfcOpenShell.Net/IfcOpenShell.Net/IfcFileExtensions.cs

```diff
@@ -0,0 +1,55 @@
+using System;
+
+namespace IfcOpenShell
+{
+    public static class IfcFileExtensions
+    {
+
+        /// <summary>
+        /// Returns a unit scale factor to convert to and from IFC project units and SI units.
+        /// </summary>
+        /// <returns>The scale factor</returns>
+        public static double CalculateUnitScale(this IfcFile file, string unitType = "LENGTHUNIT")
+        {
+            var unitAssignment = file.ByType("IfcUnitAssignment");
+            if (unitAssignment == null || unitAssignment.Count == 0)
+            {
+                return 1.0;
+            }
+            var units = unitAssignment[0];
+            var unitScale = 1.0;
+
+            if (units.TryGetAttributeAsEntityList("Units", out var unitList))
+            {
+                foreach (var unit in unitList)
+                {
+                    if (!unit.TryGetAttributeAsString("UnitType", out var localUnitType) ||
+                        !localUnitType.Equals(unitType))
+                    {
+                        continue;
+                    }
+                    var localUnit = unit;
+                    while (localUnit != null && localUnit.Is("IfcConversionBasedUnit"))
+                    {
+                        if (localUnit.TryGetAttributeAsEntity("ConversionFactor", out var convFactor))
+                        {
+                            unitScale *= convFactor.TryGetAttributeAsDouble("ValueComponent", out var convValue) ? convValue : 1.0;
+                            localUnit = convFactor.TryGetAttributeAsEntity("UnitComponent", out var unitComp) ? unitComp : null;
+                        }
+                    }
```

Comment:
When thinking about an implementation of a method that tries to get a unit assignment, i did not expect find a nested while loop.
Can you add a comment that explains the necessity for such a complex search, given the complexity of the IFC standard? Feel free to add a link to the ifc converstion based unit